library/lua/gui/widgets.lua

```diff
@@ -2,3338 +2,40 @@
-
-    return math.abs(mouse_pos.x - anchor_pos.x) + 1,
-        math.abs(mouse_pos.y - anchor_pos.y) + 1,
-        math.abs(mouse_pos.z - anchor_pos.z) + 1
-end
-
-function DimensionsTooltip:init()
-    ensure_key(self, 'frame').w = 17
-    self.frame.h = 4
-
-    self.label = Label{
-        frame={t=0},
-        auto_width=true,
-        text={
-            {
-                text=function()
-                    local anchor_pos = self.get_anchor_pos_fn()
-                    return ('%dx%dx%d'):format(get_cur_area_dims(anchor_pos))
-                end
-            }
-        },
-    }
-
-    self:addviews{
-        Panel{
-            -- set minimum size for tooltip frame so the DFHack frame badge fits
-            frame={t=0, l=0, w=7, h=2},
-        },
-        self.label,
-    }
-end
-
-function DimensionsTooltip:render(dc)
-    local x, y = dfhack.screen.getMousePos()
-    if not x or not self.get_anchor_pos_fn() then return end
-    local sw, sh = dfhack.screen.getWindowSize()
-    local frame_width = math.max(9, self.label:getTextWidth() + 2)
-    self.frame.l = math.min(x + self.display_offset.x, sw - frame_width)
-    self.frame.t = math.min(y + self.display_offset.y, sh - self.frame.h)
-    self:updateLayout()
-    DimensionsTooltip.super.render(self, dc)
-end
+Widget = require('gui.widgets.widget')
+Divider = require('gui.widgets.divider')
+Panel = require('gui.widgets.containers.panel')
+Window = require('gui.widgets.containers.window')
+ResizingPanel = require('gui.widgets.containers.resizing_panel')
+Pages = require('gui.widgets.containers.pages')
+EditField = require('gui.widgets.edit_field')
+HotkeyLabel = require('gui.widgets.labels.hotkey_label')
+Label = require('gui.widgets.labels.label')
+Scrollbar = require('gui.widgets.scrollbar')
+WrappedLabel = require('gui.widgets.labels.wrapped_label')
+TooltipLabel = require('gui.widgets.labels.tooltip_label')
+HelpButton = require('gui.widgets.buttons.help_button')
+ConfigureButton = require('gui.widgets.buttons.configure_button')
+BannerPanel = require('gui.widgets.containers.banner_panel')
+TextButton = require('gui.widgets.buttons.text_button')
+CycleHotkeyLabel = require('gui.widgets.labels.cycle_hotkey_label')
+ButtonGroup = require('gui.widgets.buttons.button_group')
+ToggleHotkeyLabel = require('gui.widgets.labels.toggle_hotkey_label')
+List = require('gui.widgets.list')
+FilteredList = require('gui.widgets.filtered_list')
+TabBar = require('gui.widgets.containers.tab_bar')
+RangeSlider = require('gui.widgets.range_slider')
+DimensionsTooltip = require('gui.widgets.dimensions_tooltip')
+
+Tab = TabBar.Tab
+makeButtonLabelText = Label.makeButtonLabelText
+parse_label_text = Label.parse_label_text
+render_text = Label.render_text
+check_text_keys = Label.check_text_keys
+
+DOUBLE_CLICK_MS = Panel.DOUBLE_CLICK_MS
+STANDARDSCROLL = Scrollbar.STANDARDSCROLL
+SCROLL_INITIAL_DELAY_MS = Scrollbar.SCROLL_INITIAL_DELAY_MS
+SCROLL_DELAY_MS = Scrollbar.SCROLL_DELAY_MS
```

Comment:
we'll probably need a function-based API for these config elements. As written, if you change these values via the control panel, the relevant widgets won't actually see the changed value and will not change their behavior. See `scripts/internal/control-panel/registry.lua` for how these values are interacted with.

We could change them to a function interface, like we did for `HIDE_ARMOK_TOOLS` in `dfhack.lua`. It is now wrapped with `getMortalMode()` and `setMortalMode()` functions. The code in the control panel registry would then have to be updated to call the new API instead of setting the values directly.